Step3 - 로또(자동)

lsm7179 opened the pull request:
안녕하세요.
리뷰어님 코드리뷰 부탁드립니다.

두서없이 한거같아서 부족한 부분이 많습니다. ㅜㅜ
잘 부탁드립니다. 감사합니다.

lsm7179 committed:
refactor: 매직넘버 상수로 변경

lsm7179 committed:
docs: 3단계-로또 요구사항 작성

요구사항 및 구현 예시 작성

lsm7179 committed:
feat: 로또 객체를 생성하는 기능 구현

lsm7179 committed:
feat: 로또를 자동 생성하는 기능 구현

lsm7179 committed:
feat: 숫자를 입력하면 입력한 갯수만큼 로또를 생성하는 기능

lsm7179 committed:
feat(PriceUtil): 금액을 나눠서 로또를 몇개 사는지 계산하는 기능

lsm7179 committed:
feat(View): View 클래스 생성

lsm7179 committed:
feat(WinningLotto): 당첨 번호를 입력받아 당첨로또 번호 객체 생성하는 기능

lsm7179 committed:
feat(WinningLotto): 몇개가 일치 했는지 확인하는 기능

lsm7179 committed:
feat: 당첨 번호가 몇개가 일치 했는지 확인하는 기능

lsm7179 committed:
refactor(LottoAutoCreateFactory): 일급 콜렉션 사용

lsm7179 committed:
feat: 로또 컨트롤러 생성

lsm7179 committed:
feat: 로또숫자 범위에 맞게 생성하고 범위가 아니면 에러나는 기능 추가

lsm7179 committed:
refactor: lottoNumber 원시값 포장하여 리팩토링

lsm7179 committed:
chore : Lotto.java model 패키지로 이동

lsm7179 committed:
feat: 당첨 값 상태를 가지는 객체 생성

lsm7179 committed:
refactor: 당첨리턴 원시값을 포장하는 객체로 변경

lsm7179 committed:
feat: 로또 당첨 상태를 가지는 기능 추가

lsm7179 committed:
feat: "당첨 통계 결과 및 수익률 결과 기능 추가

lsm7179 committed:
fix: 수익률 결과 안나오는 버그 수정

lsm7179 committed:
feat(View): 수익률 출력 기능

lsm7179 committed:
feat(creatLotto): 문자열을 입력받아 당첨 로또를 생성하는 기능 추가

lsm7179 committed:
feat: 로또 프로그램 기능 구현

lsm7179 committed:
refactor: 매직넘버 정리

lsm7179 committed:
refactor: 불필요한 임포트 정리, 매직넘버 정리

lsm7179 committed:
feat: 구입금액 에러 처리 기능 추가

lsm7179 committed:
feat: 문자열을 나눠서 int 배열로 만드는 기능

lsm7179 committed:
feat: 당첨 번호를 입력하는 기능 에러 처리

lsm7179 committed:
feat: 입력 받은 숫자 배열이 중복이면 에러가 나오게 하는 기능 추가

lsm7179 committed:
style: 코드 포맷팅

lsm7179 committed:
docs: 불필요한 내용 제거

kang-hyungu commented on src/main/java/lotto/controller/LottoController.java:
```diff
@@ -0,0 +1,84 @@
+
+import java.math.BigInteger;
+
+public class LottoController {
```
컨트롤러가 무엇일까요?
컨틀롤러가 왜 필요할까요?

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,47 @@
+    private List<LottoNumber> lottoNumbers;
+
+    public Lotto(int[] numbers) {
+        valid(numbers.length);
```
numbers가 null인 경우는 없을까요?

kang-hyungu commented on src/main/java/lotto/model/LottoNumber.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.Objects;
+
+public class LottoNumber {
+    private final int number;
```
상수를 먼저 작성해주세요.
구글 자바 코딩 스타일 참고하시면 됩니다.

kang-hyungu commented on src/main/java/lotto/model/LottoNumber.java:
```diff
@@ -0,0 +1,41 @@
+    private final int number;
+    public static final int MIN_NUMBER = 1;
+    public static final int MAX_NUMBER = 45;
+    public static final int SIZE = 6;
```
로또 번호가 전체 로또 갯수를 관리할 필요가 있을까요?

kang-hyungu commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -0,0 +1,44 @@
+
+public class Lottos {
+
+    private final List<Lotto> lottoList;
```
타입이 List인데 변수명에 중복해서 List로 표기할 필요가 있을까요?

kang-hyungu commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -0,0 +1,44 @@
+        return result;
+    }
+
+    private Map<Rank, Integer> createResultMap() {
```
로또 결과를 처리하는 객체를 추가하면 어떨까요?

kang-hyungu replied:
Result가 있군요?
Result에게 Lottos와 당첨번호를 파라미터로 넘겨서 계산하도록 맡겨도 되지 않을까요?

lsm7179 committed:
refactor: 리뷰 반영

numbers가 null인 경우처리
코드스타일 상수 먼저 작성
로또 갯수 Lottos가 관리하도록 변경
변수명에 타입을 적지 않도록 변경
Result에게 Lottos와 당첨번호를 파라미터로 넘겨서 통계 계산하도록 변경
InputHandler View 패키지로 변경
게임 규칙을 도메인 객체에서 관리하도록 변경
PriceUtil -> Price로 변경

kang-hyungu commented on src/main/java/lotto/util/InputHandler.java:
```diff
@@ -0,0 +1,55 @@
+package lotto.util;
```
왜 util 패키지 인가요?
입출력과 관련된 클래스 아닌가요?

kang-hyungu commented on src/main/java/lotto/util/InputHandler.java:
```diff
@@ -0,0 +1,55 @@
+import lotto.view.ErrorMessage;
+
+public class InputHandler {
+    private static final int LOTTO_MAX_BUY_LENGTH = 6;
```
로또 번호 최대 갯수를 로또 도메인과 입력 클래스에서 따로 관리하고 있네요.
게임 규칙이 5자리로 바뀌면 두 곳을 모두 수정해야 합니다.
실수로 한 곳만 수정하게 되면 버그가 되겠네요.
게임 규칙이니 도메인 객체에서 관리하도록 수정하는게 좋겠네요.

kang-hyungu commented on src/main/java/lotto/util/PriceUtil.java:
```diff
@@ -0,0 +1,16 @@
+
+import java.math.BigInteger;
+
+public class PriceUtil {
```
왜 util을 붙이셨나요?
Price 도 도메인으로 봐도 되지 않나요?

kang-hyungu reviewed (changes requested):
승민님 몇 가지 피드백 남겼습니다.
수정해보시고 리뷰 재요청주세요~!

lsm7179 committed:
refactor: 리뷰 반영

Controller 삭제

lsm7179 committed:
fix: 통계안되는 버그 수정

kang-hyungu commented on src/main/java/Application.java:
```diff
@@ -0,0 +1,50 @@
+import lotto.view.View;
+
+
+public class Application {
```
메인 클래스가 너무 복잡합니다.
아래와 같은 형태로 바꾸고 메인 클래스에 있는 메서드들은 적절한 역할을 맡은 클래스들로 옮겨보아요

```java
public class Application {
    public static void main(String[] args) {
        LottoGame lottoGame = new LottoGame();
        lottoGame.run();
}

```

kang-hyungu commented on src/main/java/lotto/service/LottoCreateFactory.java:
```diff
@@ -0,0 +1,61 @@
+package lotto.service;
```
팩토리 클래스를 서비스 객체로 보는게 적절할까요?

kang-hyungu commented on src/main/java/lotto/model/Price.java:
```diff
@@ -0,0 +1,16 @@
+import java.math.BigInteger;
+
+public class Price {
+    private static final int lottoPrice = 1000;
```
자바 코딩 컨벤션을 맞춰주세요

kang-hyungu commented on src/main/java/lotto/service/LottoCreateFactory.java:
```diff
@@ -0,0 +1,61 @@
+    private static int[] createRandomNumbers() {
+        List<Integer> numbers = makeNumbers();
+        Collections.shuffle(numbers);
+        int[] result = new int[Lotto.SIZE];
```
List 기능을 사용하면 배열에 값을 직접 담는 코드를 작성하지 않아도 됩니다.
subList 를 활용해보시고 for문과 배열로 반환하는 코드를 수정해주세요.
배열 보다 List 사용을 권장합니다.

kang-hyungu commented on src/main/java/lotto/model/Price.java:
```diff
@@ -0,0 +1,16 @@
+public class Price {
+    private static final int lottoPrice = 1000;
+
+    public static int getCount(int price) {
```
메서드명이 get으로 시작하는게 적절할까요?
이 메서드는 어떤 기능을 수행할까요?

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+    public static final int SIZE = 6;
+    private List<LottoNumber> lottoNumbers;
+
+    public Lotto(int[] numbers) {
```
전체 코드에서 배열을 리스트로 변경해주세요.
자바에서 제공하는 컬렉션의 다양한 기능을 사용할 수 있고 배열의 인덱스 때문에 발생하는 이슈들을 피할 수 있습니다.

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+        return lottoNumbers.size();
+    }
+
+    public boolean compare(LottoNumber lottoNumber) {
```
내부에서만 사용하는 메서드이니 public으로 열어둘 필요가 없어보입니다.

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+
+public class Lotto {
+    public static final int SIZE = 6;
+    private List<LottoNumber> lottoNumbers;
```
```suggestion
    private final List<LottoNumber> lottoNumbers;
```

kang-hyungu commented on src/main/java/lotto/model/Lotto.java:
```diff
@@ -0,0 +1,51 @@
+    public Lotto(int[] numbers) {
+        valid(numbers);
+        lottoNumbers = new ArrayList<>();
+        for (int i = 0; i < numbers.length; i++) {
```
foreach문으로 바꾸면 가독성이 높아지겠네요

kang-hyungu commented on src/main/java/lotto/model/Result.java:
```diff
@@ -0,0 +1,63 @@
+    public String toString() {
+        StringBuilder sb = new StringBuilder("");
+        matchResult.forEach((rank, count) -> {
+            sb.append(String.format(Message.STATS, rank.getCountOfMatch(), rank.getWinningMoney(), count));
```
모델 객체가 뷰 객체를 사용하면 안 됩니다.
콘솔이 아닌 모바일, 웹으로 확장하면 이 메서드는 재사용 못합니다.
뷰 -> 모델 흐름이 되도록 바꿔주세요.

kang-hyungu reviewed (changes requested):
승민님 추가 피드백 남겼습니다.
수정해보시고 리뷰 재요청주시고 궁금한 점은 코멘트 남겨주시거나 DM 주세요

lsm7179 committed:
refactor: 리뷰사항 반영

lsm7179 committed:
refactor: 통계결과 문자열 ResultView로 이동

lsm7179 committed:
refactor: 수익률 계산 리팩토링

lsm7179 commented:
안녕하세요~!
강현구님 매번 리뷰 꼼꼼하게 해주셔서 감사합니다.
리뷰 사항 반영해서 요청드립니다.
감사합니다.

lsm7179 committed:
refator : foreach 사용

lsm7179 committed:
refator : list를 set으로 변환하는 기능 리팩토링

lsm7179 committed:
refactor: foreach 로 변경

kang-hyungu commented on src/main/java/lotto/LottoGame.java:
```diff
@@ -0,0 +1,53 @@
+    private static Lottos lottos;
+    private static Lotto winLotto;
+
+    public void run() {
```
static 변수를 제거하고 아래와 같이 코드를 바꿔주세요
특별한 이유가 없다면 static 변수는 사용하지 않아야 합니다.

```java
    public void run() {
        try {
            final Lottos lottos = buyLotto();
            final Lotto winningLotto = inputWinningLotto();
            statistics(lottos, winningLotto);
        } catch (Exception e) {
            View.print(e.getMessage());
        }
    }
```

kang-hyungu commented on src/main/java/lotto/factory/LottoCreateFactory.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public static Lottos createLottos(int size) {
+        Lotto[] lottoArray = new Lotto[size];
```
배열 대신 리스트로 변경해주세요.

kang-hyungu commented on src/main/java/lotto/model/Lottos.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    @Override
+    public String toString() {
```
모델 객체에서 toString 메서드가 뷰 객체에서 사용하기 위해 변형되었네요.
toString을 제거하시고 뷰 -> 모델 흐름이 되도록 바꿔주세요.
뷰에 출력하기 위해 필요하다면 getter를 최소화해서 사용해도 좋습니다.

kang-hyungu commented on src/test/java/lotto/model/LottosTest.java:
```diff
@@ -0,0 +1,46 @@
+
+    @DisplayName("수익률 결과")
+    @Test
+    void name() {
```
1. 메서드명을 적절히 바꿔보아요
2. 이 테스트는 Result 객체의 기능을 확인하는 테스트네요. ResultTest를 추가해서 코드를 옮겨보아요

kang-hyungu reviewed (approved):
승민님 피드백 반영 잘 해주셨어요
추가 코멘트 남겼는데 다음 미션 진행하면서 적용해주세요
pr은 머지 처리하겠습니다. 수고하셨어요
